### PR TITLE
fetcher: Add a maximum layer size

### PR DESCRIPTION
**Pull request by crozzy**
This change adds a limit when downloading the layer files. The layer size is computed from the content-length reported from a HEAD request, this means it'll often be the size of the compressed layer and it requires the storage to accurately report the content-length.

Signed-off-by: crozzy <joseph.crosland@gmail.com>

### CODE REVIEW

File: libindex/fetcher.go [hunk 1]
```diff
@@ -206,14 +209,32 @@ func (a *RemoteFetchArena) realizeLayer(ctx context.Context, l *claircore.Layer)
...
+		case err == nil:
+			if a.maxBytes > 0 && size > a.maxBytes {
+				return "", fmt.Errorf("fetcher: layer size (%d) larger than maximum layer size (%d)", size, a.maxBytes)
+			}
+		default:
+			zlog.Warn(ctx).Msg("unable to convert content-length header to bytes")
+		}
+	}
+
+	req.Method = http.MethodGet
```
Comment (hdonnay) [inline libindex/fetcher.go:233]: This won't work with signed URLs, they usually have only a set of allowed methods.

Comment (crozzy) [reply libindex/fetcher.go]: I changed it to just make the GET request which means we should avoid any latency implication as long as the length check comes before the body reading.

Comment (crozzy) [reply libindex/fetcher.go]: need to update the commit msg

**Comment by hdonnay** [comment]
This isn't the maximum layer size, though. It's a transfer limit, as we currently need to decompress it.

As implemented, it would make a configuration hard to reason about: if an operator allows 10 layers with a "maximum layer size" of of 500 MiB then allocates 5000MiB of disk, it's still possible (and expected behavior) to use more disk than that.

**Comment by crozzy** [comment]
My thinking is that if you have a registry that allows layer files big enough that the request to Clair times out before it can even download all the compressed bytes then we should have a knob to configure a reasonable upper bound. The naming might be a bit off but essentially it's not about the storage but about the transfer time. The caller checking the size is also an option.

**Comment by hdonnay** [comment]
If it's about a timeout, then the knob should be expressed in time.

**Comment by crozzy** [comment]
Right, the point was to get out early before doing work, but I think we're on the same page that perhaps this would be more trouble to explain than being useful.

**Pull request closed without merging**